// @material-ui-molecule/jss.js
// import _extends from "../@babel/runtime/helpers/builtin/extends.js";
// import * as mui from '../@material-ui/core/';
import _objectWithoutProperties from "../@babel/runtime/helpers/builtin/objectWithoutProperties.js";
import { fade } from '../@material-ui-molecule/core/styles/colorManipulator.js';
import { capitalize } from '../@material-ui-molecule/core/utils/helpers.js';
import _objectSpread from  "../@babel/runtime/helpers/builtin/objectSpread.js";
import warning from '../warning/browser.js';
import createMuiTheme from '../@material-ui-molecule/core/styles/createMuiTheme.js'
import deepmerge from '../deepmerge/dist/umd.js'; 
import createGenerateClassName from '../@material-ui-molecule/core/styles/createGenerateClassName.js';
import {Molecule} from '../molecule/molecule-comp.js';

import "../jss/dist/jss.js"
import "../jss-global/dist/jss-global.js"
import "../jss-nested/dist/jss-nested.js"
import "../jss-camel-case/dist/jss-camel-case.js"
import "../jss-default-unit/dist/jss-default-unit.js"
import "../jss-vendor-prefixer/dist/jss-vendor-prefixer.js"
import "../jss-props-sort/dist/jss-props-sort.js"
import "../classnames/index.js"
export const defaultTheme = createMuiTheme();

export const themes = window.THEMES || {}

themes['default'] = defaultTheme;
defaultTheme.name = 'default';

function jssPreset() {
    return {
        plugins: [jssGlobal.default(), jssNested.default(), jssCamelCase.default(), jssDefaultUnit.default(), jssVendorPrefixer.default(), jssPropsSort.default()]
    };
}

jss.default.setup(jssPreset())

const generateClassName = createGenerateClassName();
const sheetOptions = {generateClassName}
const stylesCreatorSavedOptions = {index:-99999999997}

const loadedClasses = {};

export function attachJss(styles, meta, name, theme=defaultTheme){
    const key = `${theme.name}-${meta}-${name}`;
    const exist = loadedClasses[key]
    if(exist) return exist;

    if(typeof styles == 'function'){
        styles = styles(theme);
    }
    const sheet = jss.default.createStyleSheet(styles, _objectSpread({
        meta,
        classNamePrefix: meta,
        flip: theme.direction === 'rtl',
        link: false
    }, sheetOptions, stylesCreatorSavedOptions, {
        name
    }));
    sheet.attach()
    
    loadedClasses[key] = sheet.classes;

    return sheet.classes
};

Molecule.prototype.attachJss = function(styles, meta, name){
    const p = this.element.closest('[theme]');
    const themeName = p ? p.getAttribute('theme') : 'default';
    const theme = themes[themeName] || defaultTheme;
    return attachJss(styles, meta || this.constructor.name, name || this.constructor.name, theme);
}
